libvast/src/evaluate.cpp

```diff
@@ -0,0 +1,465 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2022 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "vast/arrow_table_slice.hpp"
+#include "vast/bitmap_algorithms.hpp"
+#include "vast/detail/assert.hpp"
+#include "vast/detail/overload.hpp"
+#include "vast/detail/passthrough.hpp"
+#include "vast/expression.hpp"
+#include "vast/ids.hpp"
+#include "vast/logger.hpp"
+#include "vast/table_slice.hpp"
+#include "vast/type.hpp"
+
+#include <arrow/record_batch.h>
+
+#include <cstddef>
+#include <span>
+
+namespace vast {
+
+namespace {
+
+// A utility function for evaluating meta extractors in predicates. This is
+// always a yes or no question per batch, so the function does not have to deal
+// with bitmaps at all.
+bool evaluate_meta_extractor(const table_slice& slice,
+                             const meta_extractor& lhs, relational_operator op,
+                             const data& rhs) {
+  switch (lhs.kind) {
+    case meta_extractor::kind::type:
+      return evaluate(materialize(slice.layout().name()), op, rhs);
```

Comment:
Is materialization needed here?

Reply:
We did not touch this code; we could probably delegate to the cell_evaluator here instead, then we wouldn't need to materialize at all.

Reply:
I did end up pushing a commit that implements this in terms of the cell evaluator. It's a lot more verbose, but hey, it's one step further towards not needing `evaluate(data, op, data) -> bool` at all anymore.